scanpy/neighbors/__init__.py

```diff
@@ -111,6 +112,66 @@ def neighbors(
     return adata if copy else None
 
 
+def _rp_forest_generate(rp_forest_dict):
+    from collections import namedtuple
+
+    props = ['hyperplanes', 'offsets', 'children', 'indices']
+    FlatTree = namedtuple('FlatTree', props)
+    num_trees = len(rp_forest_dict[props[0]]['start'])-1
+
+    for i in range(num_trees):
+        tree = []
+        for prop in props:
+            start = rp_forest_dict[prop]['start'][i]
+            end = rp_forest_dict[prop]['start'][i+1]
+            tree.append(rp_forest_dict[prop]['data'][start:end])
+        yield FlatTree(*tree)
+
+    tree = []
+    for prop in props:
+        start = rp_forest_dict[prop]['start'][num_trees]
+        tree.append(rp_forest_dict[prop]['data'][start:])
+    yield FlatTree(*tree)
+
+
+def neighbors_update(adata, adata_new, k=10, queue_size=5, random_state=0):
+    # only with use_rep='X' for now
+    from umap.nndescent import make_initialisations, make_initialized_nnd_search, initialise_search
+    from umap.umap_ import INT32_MAX, INT32_MIN
+    from umap.utils import deheap_sort
+    import umap.distances as dist
+
+    if 'metric_kwds' in adata.uns['neighbors']['params']:
+        dist_args = tuple(adata.uns['neighbors']['params']['metric_kwds'].values())
+    else:
+        dist_args = ()
+    dist_func = dist.named_distances[adata.uns['neighbors']['params']['metric']]
+
+    random_init, tree_init = make_initialisations(dist_func, dist_args)
+    search = make_initialized_nnd_search(dist_func, dist_args)
+
+    search_graph = adata.uns['neighbors']['distances'].copy()
+    search_graph.data = (search_graph.data > 0).astype(np.int8)
+    search_graph = search_graph.maximum(search_graph.transpose())
+    # prune it?
+
+    random_state = check_random_state(random_state)
+    rng_state = random_state.randint(INT32_MIN, INT32_MAX, 3).astype(np.int64)
+
+    if 'rp_forest' in adata.uns['neighbors']:
+        rp_forest = _rp_forest_generate(adata.uns['neighbors']['rp_forest'])
+    else:
+        rp_forest = None
+    train = adata.X
+    test = adata_new.X
+
+    init = initialise_search(rp_forest, train, test, int(k * queue_size), random_init, tree_init, rng_state)
+    result = search(train, search_graph.indptr, search_graph.indices, init, test)
+
+    indices, dists = deheap_sort(result)
+    return indices[:, :k], dists[:, :k]
```

Comment:
Just checking, since downstream functions like `leiden` use the `"connectivities"` weighted graph, are you planning on returning that as well?

Reply:
For now this is just an analog of
```
neighbors = NNDescent(adata1.X)  # construct first graph
neighbors.query(adata2.X)  # lookup second data
```
But calculating `connectivities` should be possible, yes.